### PR TITLE
Overhaul GameParser

### PR DESCRIPTION
**Pull request by RoiEXLab**
Follow-up of #1883

### CODE REVIEW

**Comment by RoiEXLab** [comment]
@ron-murhammer @DanVanAtta Tests are fixed, ready for review

File: src/main/java/games/strategy/engine/data/GameParser.java [hunk 1]
```diff
@@ -447,14 +406,14 @@ private Object getInstance(final String className) throws GameParseException {
...
     } catch (final InstantiationException e) {
       throw new GameParseException(mapName,
-          "Class <" + className + "> could not be instantiated. ->" + e.getMessage());
+          "Class <" + className + "> could not be instantiated", e);
     } catch (final IllegalAccessException e) {
-      throw new GameParseException(mapName, "Constructor could not be accessed ->" + e.getMessage());
+      throw new GameParseException(mapName, "Constructor could not be accessed", e);
     } catch (ReflectiveOperationException e) {
-      throw new GameParseException(mapName, "Exception while constructing object ->" + e.getMessage());
+      throw new GameParseException(mapName, "Exception while constructing object", e);
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/GameParser.java]: Now that you are including the cause in `GameParseException`, I don't think it's necessary to have different messages for different types of exceptions.  You should be able to reduce this entire method to the following:

```java
private Object getInstance(final String className) throws GameParseException {
  try {
    return Class.forName(className).getDeclaredConstructor().newInstance();
  } catch (final ReflectiveOperationException e) {
    throw new GameParseException(mapName, String.format("Unable to create instance of class <%s>", className), e);
  }
}
```

File: src/main/java/games/strategy/engine/data/GameParser.java [hunk 2]
```diff
@@ -298,142 +305,94 @@ private PlayerID getPlayerID(final Element element, final String attribute, fina
...
-      throw new GameParseException(mapName, "Could not find technology. name:" + name);
-    }
-    return type;
+    return getValidatedObject(element, attribute, mustFind, name -> {
+      TechAdvance type = data.getTechnologyFrontier().getAdvanceByName(name);
+      if (type == null) {
+        type = data.getTechnologyFrontier().getAdvanceByProperty(name);
+      }
+      return type;
+    }, "technology");
```
Comment (ron-murhammer) [inline src/main/java/games/strategy/engine/data/GameParser.java]: Is there any cleaner way to write this one? This one is a little difficult to read.

Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/GameParser.java]: Can't think of a better way... Doesn't seem too complicated to me \:-/

**Comment by RoiEXLab** [comment]
@ssoloff @ron-murhammer Updated!

File: src/main/java/games/strategy/engine/data/GameParser.java (hunk 2 above)
Comment (ron-murhammer) [inline src/main/java/games/strategy/engine/data/GameParser.java]: I think the part I don't like is the function declaration in the middle of listing method parameters. Maybe defining the function first then passing it to the method?

Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/GameParser.java]: But I prefer avoiding local variables that are only used once...

Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/data/GameParser.java]: Some quick suggestions:


Use `type` variable just once:

```
     TechAdvance type = data.getTechnologyFrontier().getAdvanceByName(name);
     if (type != null) {
       return type;
     }
     return data.getTechnologyFrontier().getAdvanceByProperty(name);
```


Extract to a method:
```

     return techAdvanceType(data.getTechnologyFrontier());
    }, "technology");
    :
    :


    private static TechAdvance techAdvanceType( ...) {
      TechAdvanceType type = data.getTechnologyFrontier().getAdvanceByName(name);
       if (type != null) {
         return type;
       }
       return data.getTechnologyFrontier().getAdvanceByProperty(name);
    }
```


Optional route:
```
   return Optional.ofNullable(data.getTechnologyFrontier().getAdvanceByName(name))
              .orElse(data.getTechnologyFrontier().getAdvanceByProperty(name));
```

Other options would likely be slightly more involved, but perhaps better, notably updating technology frontier so clients do not need to worry about this.

**Comment by ssoloff** [comment]
@RoiEXLab The exception changes look good.  However, I kind of agree with the suggestion to extract the technology frontier lambda to a method or a local variable with a meaningful name.  All the other refactored methods read very nicely.

**Comment by RoiEXLab** [comment]
@DanVanAtta @ron-murhammer @ssoloff Updated!

**Pull request merged**